PvE Gamemode, Skillcards, Trading, Gifting, Tagplay gamemode, Daily login bonus & Game invites

YestinJarrett opened the pull request:
- **All PvE packets** have been reversed and implemented, working NPC enemies that make use of the A* pathfinding algorithm to navigate around. All NPC enemies and drops have been implemented, as well as the respawn system upon death using gold coins. First and second level/map of PvE have been fully implemented as well as the 2 difficulties that come with it and are fully playable. Third map works but doesn't have any npc's coded into it, only the boss spawns. This level could be implemented in the future though. Ending the game or going to the next round works as well (with a lot of thanks to and appreciated help and insights from @hinnie123).

  To get PvE to work fully, we did have to perform quite some patches, so in order to play the PvE gamemode one would need the [Qpang Tweak Tool](https://github.com/hinnie123/qpang-tweak-tool) made by @hinnie123. This tool/dll also contains a bunch of patches to have Qpang support all resolutions (no longer limited to 1280x720) 👌

  - See [Qpang PvE/StO - Teaser](https://www.youtube.com/watch?v=wj8Vne2hs_A) and [Qpang PvE/StO - Update #1](https://www.youtube.com/watch?v=dxPE1-Dx7OM)
  

- **All skillcards** have been implemented into the game, can be bought from shop, equipped in slots and used in game, uses go down. I think we might have even added support for custom skillcards.
  - See [QFighter/Qpang - Introducing Skillcards (WIP)](https://www.youtube.com/watch?v=jtJMOhNaRlo)
- Trading has been (fully) implemented. Players can now trade items with each other.
  - See [Qpang 2021 - Trading is here!](https://www.youtube.com/watch?v=b38bN5kt0h4)

- Tagplay gamemode has been fully implemented. Game start a new "Tag" player is chosen, upon death, a new player is chosen as well. Tag player transforms into batteryman.

  - See [QFighter 2021 - Public Enemy (Tag/Prey) Mode.](https://www.youtube.com/watch?v=hnp78dBAzX8) and [QFighter 2021 - QPang Public Enemy/TagPlay Gameplay - Ep 1](https://www.youtube.com/watch?v=O_XebLCi05Q)

- Daily login bonus has been fully implemented as well as game invites, no video of this.

We also reworked the inventory system where, items are now sent in chunks to the client allowing for pretty much infinite inventory space as well as gifts. The inventory is no longer limited to 200 items. A lot of thanks to Dennis for that!

Probably a bunch more added/fixed/updated but I don't know anymore, been a while 😅

Deluze commented:
Looks good, I'll test when I get home.

Deluze commented:
Seems to work fine. I applied some formatting since I think having static declarations on top is better.

YestinJarrett committed:
Set default value to false.

YestinJarrett committed:
Started reversing CGPvEHitNpcToObject and GCPvEHitNpcToObject.

YestinJarrett committed:
Remove include.

YestinJarrett committed:
Started keeping track of coin count to show in the result screen.

YestinJarrett committed:
Set default values for coin count and fixed floorNumber type.

YestinJarrett committed:
Added time left in millis for the result screen.

Knyntsje committed:
Respawn players instead of resetting health when new PVE round starts

Knyntsje committed:
Update GCPvEHitNpc packet; hasTargetDied is actually killStreak

Knyntsje committed:
Update GCPvEHitNpcData; hasTargetDied is killStreak

Knyntsje committed:
update onCGPvEHitNpc: increase killstreak on kill, and relay the killstreak to player

Knyntsje committed:
Clamp killstreak to 255 as it's an uint8_t

Knyntsje committed:
Don't damage other players in PVE

Knyntsje committed:
Add PVE gamemode check before looping pve areas

Knyntsje committed:
Also check for current area in cg_move

YestinJarrett committed:
Added 10% damage reduction to npcs. Moved ternary operator for determining killstreak up and into its own variable for more readability imo.

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Added pve manager to the reload command.

Knyntsje committed:
Fixed aitime, needed a conversion to int

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Add canShoot to cleanup npc code

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Update test npc

Knyntsje committed:
Update above floor grid for stage 1

Knyntsje committed:
Last position of path was never being set, fixes melee npcs

Knyntsje committed:
Change endRound to finish for now

YestinJarrett committed:
Moved pve check.

YestinJarrett committed:
Commented out areas loggings.

Knyntsje committed:
Add name checks in cg_room for now

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Let RoomSession::spawnPlayer use map from roundmanager if the gamemode is pve

Knyntsje committed:
Add m_isPublicPveRoom and methods to Room

Knyntsje committed:
In CGRoom, when joining check if pve room is public

Knyntsje committed:
Add PveMakePublicCommand

Knyntsje committed:
Woops, my bad

Knyntsje committed:
Update DebugCommand to support printing grids over 60 in size

Knyntsje committed:
Pathfinder: use smart ptr, and make cache 85 * 85

Knyntsje committed:
Initialize stage 2 map info in round manager, initialize objects etc after sending newround packet

Knyntsje committed:
Update move report for stage 2 map info

Knyntsje committed:
Add back Pathfinder::free function, as round manager uses it

Knyntsje committed:
Add stage 2 map info

Knyntsje committed:
Fix object initializing bug

YestinJarrett committed:
We figured out the last unknown, which is the initial health of the object.

Knyntsje committed:
GCPvEHitNpcToObject: unk_02 is remainingHealth

Knyntsje committed:
GCPvEHitNpc: unk_21 is remainingHealth (used for boss)

Knyntsje committed:
Forgot to rename some to remaininghealth

Knyntsje committed:
GCPvESpecialAttack: rename unk_03 to resetAttack

YestinJarrett committed:
Add initialHealth property.

YestinJarrett committed:
Implemented the initialHealth and health properties.

YestinJarrett committed:
Health is now being passed to the object init event.

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Change unk_21 to remainingHealth in GCPvEHitNpcData

YestinJarrett committed:
Add RandomHelper, for getting a random number in a range.

YestinJarrett committed:
Built in weighted random item drops.

YestinJarrett committed:
Used RandomHelper.

YestinJarrett committed:
Started on implementation of npc hit object.

YestinJarrett committed:
Add initializeNpcWaves and getNpcWaveData as well as PveNpcWaveData for retrieving npcs per wave from the database.

YestinJarrett committed:
Add new PveNpc constructor that can be instantiated with PveNpcWaveData.

YestinJarrett committed:
Add m_pveWaveManager and getPveWaveManager and initialized the wave manager.

YestinJarrett committed:
Call PveWaveManager::tick in PlayerVsEnvironment::tick but only if the current round is 1 (stage 2).

YestinJarrett committed:
Called PveWaveManager::initializeWaves in RoomSessionPveRoundManager::onStartNewRound but only if round is 1 (stage 2).

YestinJarrett committed:
Implemented npcs spawning on an interval (or as we know it "waves").

YestinJarrett committed:
Figured out the last unknown; health.

YestinJarrett committed:
Disabled relaying damage to objects.

YestinJarrett committed:
Passed m_health to pve init.

Knyntsje committed:
Change objectUid to npcUid in CGPvEHitNpcToObject

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Added all weapon types.

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Remove validity checks from CGPvEFloor

Knyntsje committed:
Add npc essence priority logic

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Npc weapon type is now being retrieved from the database.

YestinJarrett committed:
CGPvEHitNpcToObject is now properly being relayed with the correct amount of damage being dealt per npc weapon type.

YestinJarrett committed:
Npcs now have an extra property for eWeaponType.

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Add and implement eRound enum.

Knyntsje committed:
Add PveNpc::getLastAttackerId

Knyntsje committed:
If an npc kills the player that has last attacked him, he has gotten his revenge, and the lastAttackerId will be reset to 0

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
PveNpc: Fix potential crash in attackTargetPlayer, doNextMove: call attackTargetPos if not moving to player

Knyntsje committed:
PveNpc: doNextMove, move clearPath to after attacking, so that the essence prio logic can't overwrite targetplayer in the meantime

Knyntsje committed:
Woops

YestinJarrett committed:
Add eNpcBodyPartAttributeType, eNpcBodyPartDamageType and eNpcBodyPartPartType.
Added above enums to NpcBodyPart

YestinJarrett committed:
Add method to get a body part by id.

YestinJarrett committed:
Merge changes from hinnie.

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Bodyparts now get ordered by their body part id.

YestinJarrett committed:
Implemented a better damage system based on body part.

YestinJarrett committed:
Check for body part, and pass through the body part.

Knyntsje committed:
RoomSession: add didWin arg to finishPveGame, add areAllPlayersPermanentlyDead function

Knyntsje committed:
RoomSessionPveRoundManager::checkRoundZeroFinished: call finishPveGame with true as win arg

Knyntsje committed:
CGPvERespawnReq: If all players are permanently dead, finish the game

Knyntsje committed:
Add mp as alias for PveMakePublic command

Knyntsje committed:
Also mark player as permanently dead when it's messing with packets and saying it has enough coins, but doesn't have enough coins

YestinJarrett committed:
Reworked starting a new round, the round gets started only when all players are initialized.

YestinJarrett committed:
Removed member that wasnt used.

YestinJarrett committed:
Reworked currentRound in RoundManager to be of type eRound instead of uint32_t.
Add method stopTime which pretty much sets the start time to 1 minute in the future so there is enough time between stages before it starts counting down again.

YestinJarrett committed:
Add check for elapsed time. This will make it so the area can not be triggered, only if the game has started (start countdown time has ended). This will result in npcs being spawned only if the game has started and not before that.

YestinJarrett committed:
Synchronize time for all players on round end.

YestinJarrett committed:
Add function to generate random float.

YestinJarrett committed:
The yCorrection for npc attack target position is now random between 0 and 1.

YestinJarrett committed:
Add damage modification (headshot only for now) based on bodypart hitlocation.

YestinJarrett committed:
We figured out playerId as unk_11 and remainingHealth as unk_12, unk_09 is a bool

YestinJarrett committed:
Set bool back to default

Knyntsje committed:
Switch over to storing player id instead of pointer in PveNpc

YestinJarrett committed:
Reversed GCPvEDestroyParts. This packet is for destroying/visually exploding bodyparts (in combination with body part attribute T_DESTROY).

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
The game now finishes once the essence is destroyed.

Knyntsje committed:
RoomSessionNpcManager: Add recursive mutex and lock_guard wherever m_npcs and/or m_spawnedNpcs are used

Knyntsje committed:
PveNpc: Pass npcUid to npcDoNextMove instead of npc pointer, and use findNpcByUid to get the pointer

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Removed some lockguards

YestinJarrett committed:
Only when the player is the master of the room the event gets relayed.

YestinJarrett committed:
Add back functionalitity to spawn npc.

YestinJarrett committed:
Fixed entities (such as ocotomines) not exploding for others in the room by relaying the cg_hit but setting the damage to 0 for pve.

YestinJarrett committed:
Commented out throwables not dealing damage.

YestinJarrett committed:
Make spycams explode on death, just to see if people like it otherwise revert.

Knyntsje committed:
Perhaps fix checkRoundZeroFinished crash

Knyntsje committed:
Add some mutex locks back

YestinJarrett committed:
Moved m_itemManager and m_gameMode tick methods so they get called after the player tick methods.
Add game finish check for pve in canFinish.
Add back the canFinish check in RoomSession::tick.

YestinJarrett committed:
Destroy essence object when health is at 0.

YestinJarrett committed:
Add checkRoundOneFinished to check if round 1 (stage 2) has finished.
Add check for time left in both checkRoundZeroFinished and checkRoundOneFinished.

YestinJarrett committed:
Improved check for time over.

YestinJarrett committed:
Set finish game win to true, and added todo.

YestinJarrett committed:
Added an extra check for pve for applying effects.

YestinJarrett committed:
Commented out a check for finishing the game.

YestinJarrett committed:
Add printf to SpawnNpcCommand.

Knyntsje committed:
Increase round & map on new round, not when ending round

Knyntsje committed:
Fix npc bug in stage 2

YestinJarrett committed:
Npcs in stage 2 now always drop loot.

YestinJarrett committed:
Npcs in stage 2 will now always drop loot on death.

YestinJarrett committed:
Add back checkRoundOneFinished check.
Reworked checkRoundOneFinished check and add deadPlayerCount == 0 check to prevent game from ending immediately.

YestinJarrett committed:
Moved the setCellTaken call in handleDeath to PveNpc::die, this fixes a bug where npcs could not move to cells where another npc had died which could block some pathways.

YestinJarrett committed:
Players now spawn in the correct position whilst in waiting for players in the pve gamemode.

Knyntsje committed:
Reset takenCell in PveNpc::tick, not PveNpc::die

Knyntsje committed:
Update stage 1 above ground grid: Removed some doors

YestinJarrett committed:
Hopefully, fixed a bug where pve games would end if there were 2 players and 1 player left.

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Stupid empty line i didnt notice >:(

YestinJarrett committed:
Enabled endround for stage 1 again.

Knyntsje committed:
Add note about essence logic

Knyntsje committed:
Don't let npcs focus player when they got attacked with a bomb

YestinJarrett committed:
Remove explosion on death for cameras.

YestinJarrett committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Merge movement & non-movement together. The handlers inside of them are generic enough to support that.

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
Add movement checks in PveNpc::handleTargetEssencePriority

Knyntsje committed:
Also set current cell as not taken when npc dies (don't think this can ever be taken though, but just in case)

Knyntsje committed:
Double respawn timer fix

YestinJarrett committed:
Change getPlayingPlayers to getPlayers

Knyntsje committed:
Fix new round start a bit too early

Knyntsje committed:
This respawn is not needed, it happens automatically

Knyntsje committed:
Thread safety is important

Knyntsje committed:
Change essence priority logic (npcs now don't walk to the essence first before you can lure them, revert this if we don't want it)

Knyntsje committed:
Hopefully fix 2 end round bugs

Knyntsje committed:
Added a validity check

YestinJarrett committed:
The correct extra health per pve stage is now being added (for shields that is).

YestinJarrett committed:
Add passThroughIsRequired, this will be used to check if areas were passed through.

YestinJarrett committed:
Add check in round zero finished, all required areas have to be passed through in order to finish, no cheating around this.

Knyntsje committed:
Log nullptr event in EventConnection::processEvent

Knyntsje committed:
cg_game_state: Don't call addPlayer for GAME_WAITING_PLAYERS when gamemode is pve and round is higher than stage 1

Knyntsje committed:
PveRoundManager: Call addPlayer for every player when everyone has initialized

Knyntsje committed:
Fix suicides not showing up on killfeed

Knyntsje committed:
Perhaps fix random crash

YestinJarrett committed:
Npcs from pve_npc_spawns now receive a uid from the database.

YestinJarrett committed:
The uid from the PveNpcData is now being used up on constructing an npc.

YestinJarrett committed:
Npcs with uid 0 now get assigned a uid by grabbing the lastSpawnedUid and increasing it by one, if that uid is taken it will re-try with an increased lastSpawnedUid.

YestinJarrett committed:
Merge changes from hinnie

YestinJarrett committed:
Objects will now get their uid from the uid column instead of the primary key id.

Knyntsje committed:
Update stage 1 above ground grid

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Remove db name from query.

Knyntsje committed:
Show time left also when losing in pve

Knyntsje committed:
Packets get handled in the same thread as the main tick... some mutexes can go away. Also hopefully improved stability of npcDoNextMove

Knyntsje committed:
Add stage 3 grid

Knyntsje committed:
Update pathfinder with stage 3 grid

YestinJarrett committed:
Items now have a respawn time and a bool if they should respawn or not.

YestinJarrett committed:
Implemented item respawning.

YestinJarrett committed:
Calling the tick method for pve items is something very hard to do it seems.
Moved method call spawnInitializedItems to initializeItems.

Knyntsje committed:
Add boss npc class

Knyntsje committed:
Add virtuals in PveNpc and add calculateHitDamage function

Knyntsje committed:
RoomSessionNpcManager: use PveBossNpc if the npc is of type boss

YestinJarrett committed:
Change pve_npc_body_parts.body_part_id to pve_body_parts.part_id.

Knyntsje committed:
Move around player check, perhaps fixes random finish bug

Knyntsje committed:
Add shouldExplode to GCPvENpcInit

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

Knyntsje committed:
PveBossNpc: return weapon used damage for now

Knyntsje committed:
GCMasterLog: Add constructor for GCPvESpecialAttack

Knyntsje committed:
These comments can go

YestinJarrett committed:
Npc body parts are now shared pointers.

YestinJarrett committed:
Add method to reset all body part health to initial.
Add method to take damage for a specific body part.

YestinJarrett committed:
Damage to npcs is now calculated based on body part instead of if it is a headshot or not. Broken body parts take 50% of the damage.

YestinJarrett committed:
Changed NpcBodyPart to bodyPartId.

YestinJarrett committed:
Small change due to pointer.

YestinJarrett committed:
Merge changes.

YestinJarrett committed:
Cast to uint16_t.

YestinJarrett committed:
Add frysian backpack id to the list of backpacks that give extra health.

Knyntsje committed:
Add attack members to PveBossNpc

YestinJarrett committed:
Add basic special attack start and end implementation

YestinJarrett committed:
Can not damage the boss npc whilst performing the special attack.

YestinJarrett committed:
Started working on attacking whilst in special attack.

YestinJarrett committed:
Add checkRoundTwoFinished method for checking if stage 3 (the final stage) is finished.

YestinJarrett committed:
Add override for onDeath for PveBossNpc.
Once the boss dies it will now spawn a field full of golden coins.

YestinJarrett committed:
Moved the logic for spawning the coins to it's own method.

YestinJarrett committed:
Small changes.

YestinJarrett committed:
Add map that keeps track of damage dealt by players to an npc.
Add method to update this map.

YestinJarrett committed:
Add RoomSessionBossFightManager and getter for it in the RoomSession.

YestinJarrett committed:
Moved all coin spawning logic to the boss fight manager.

YestinJarrett committed:
Using isDead instead of getHealth.. dont know why I didnt do that before but ok.

YestinJarrett committed:
The BossFightManager tick method now also gets called in stage 3.

YestinJarrett committed:
Add handler for T_DAMAGE target type. Rename m_damageDealtByPlayers to m_playerDamageToNpc.

YestinJarrett committed:
Override attackTargetPlayer and startMovingToPlayer for the BossNpc to take into account m_isInSpecialAttack. It is not perfect yet (probably because one tick method needs to be called before the other one).

YestinJarrett committed:
The PveNpc tick method is now being called after the special attack stuff, this is neccessary so the boss npc does not do one more move or attack whilst already in special attack.

YestinJarrett committed:
Add HandleOpenLootBoxRequest, response has yet to be implemented.

YestinJarrett committed:
Add todo.

YestinJarrett committed:
- Modified tick method, this way it prevents the npc from attacking whilst still on the ground when going into special attack mode, it will start attacking the next tick.

YestinJarrett committed:
If a player has not dealt damage to the npc, the player now gets added anyway, so it can be selected as a potential target.

YestinJarrett committed:
Add proper check for pve gamemode so players can not (for now) change the difficulty or pve map.

YestinJarrett committed:
Just looked some stuff up for cccharm.

YestinJarrett committed:
Re-implemented the Unlimited command to be MaxPlayers command.

YestinJarrett committed:
Figured out all bonuses for pvp and pve and reworked the GCGameState class a bit.

YestinJarrett committed:
Add some bonuses for testing if a player has won.

YestinJarrett committed:
We now spawn 4 extra bosses so we can use them to shoot in special attack.

YestinJarrett committed:
The extra boss npcs do not have to be killed, they can just be re-used.

Knyntsje committed:
Fix getRandomFloat

Knyntsje committed:
Fix boss shooting 5 bullets in special attack

Knyntsje committed:
Merge branch 'feature/pve' of https://github.com/YestinJarrett/qpang-essence-emulator into feature/pve

YestinJarrett committed:
Add override for canAttackTargetPos so we can remove the lineOfSight check for the boss only.

YestinJarrett committed:
Fixed a bug where damage to npc bodyparts would stick after games.

YestinJarrett committed:
Commented out some statements that prevent players from making pve rooms.

YestinJarrett committed:
Not sure why I did this, but it works fine so lets just keep it.

YestinJarrett committed:
Removed some print statements.

YestinJarrett committed:
Fixed correct values.

YestinJarrett committed:
Now calling resetBodyPartsHealth from constructor to fix health issue.

YestinJarrett committed:
Allow reload glitch by default for pve.

YestinJarrett committed:
Removed call npc->resetBodyPartsHealth.
Removed log.

YestinJarrett committed:
Remove logging.

YestinJarrett committed:
Fix a bug where pve rooms could be created with different starting maps.

YestinJarrett committed:
Add logging.

YestinJarrett committed:
Add forcestart command to helpers.

YestinJarrett committed:
Add some extra checks to hopefully prevent server from crashing due to effects.

YestinJarrett committed:
Some small changes, nothing special.

YestinJarrett committed:
Merge pull request #4 from YestinJarrett/feature/pve

Feature/pve

YestinJarrett committed:
Increase inventory space from 200 to 300 (fingers crossed x))

YestinJarrett committed:
Merge pull request #5 from YestinJarrett/feature/pve

Increase inventory space from 200 to 300 (fingers crossed x))

YestinJarrett committed:
Update README.md

YestinJarrett committed:
Update ConfigManager.cpp

YestinJarrett committed:
Update ConfigTemplate.h

YestinJarrett committed:
Update README.md

Deluze commented:
Wow, what a nice Christmas gift for the entire community. Thanks for opening this, really. Really nice to see many missing features be added and I'm surprised that you were able to do it in this hacked-together emulator.

I wasn't in the loop with QFighter at all, didn't know all of these things were accomplished. Nice 😄

ferib commented on src/essence.game/qpang/room/Room.cpp:
```diff
@@ -3,127 +3,162 @@
+	m_isPublicPveRoom(false)
 {
 	m_modeManager = Game::instance()->getRoomManager()->getGameModeManager()->get(mode);
+	// Reload glitch is only enabled by default in pve rooms.
```
Please note that 'reload glitch' is a skill component of the game. Disabling the reload glitch is for losers imho

ferib commented on src/essence.game/qpang/room/tnl/GameConnection.cpp:
```diff
@@ -115,6 +119,15 @@ void GameConnection::updateRoom(std::shared_ptr<Room> room, TNL::U32 cmd, TNL::U
+	postNetEvent(new GCPvEUserInit(roomPlayer));
+	postNetEvent(new GCGameState(m_player->getId(), 12));*/
+
+	startLoading(room, roomPlayer);
```
How come `GCPvEStart` and `GCPvEUserInit` weren't used? game crash? xD

ferib reviewed (approved):
Some controversial changes but il allow it xd
